### PR TITLE
Filter by types, categories and properties seperately

### PR DESCRIPTION
**Pull request by AlanSl**
- [x] https://github.com/davidmarkclements/d3-fg/pull/28
- [x] updates to tests 
- [x] update against latest master
- [x] update various things based on new v8 category from https://github.com/nearform/node-clinic-flame/pull/41
- [x] ~fixing a bug when parsing a hash that includes disabling merging~ _[update] turned out to be a bug in master, fixed in https://github.com/nearform/node-clinic-flame/pull/42_

This mostly changes things under the hood: 

 - Removes the `typeTEMP` property (long overdue!)
 - Uses isNodeExcluded() from shared.js to test for exclusion everywhere - in analysis, in the UI, and pipes it in as an override for the d3-fg test that sets `.hide` on frames
 - Therefore, exclusion tests now check for properties like isInit separately to categories like `app`, `deps`, etc, and separately to types like `v8`, `cpp`; and for deps, the types are now the name of the dependency
- Namespaces type-based exclusion by category. So, for example, if someone uses [the module `native`](https://www.npmjs.com/package/native) as a dependency, there's no ambiguity between `deps:native` and `all-v8:native` exclusion filters

This means:

- (the only visible change expected in this PR) It should be more accurate with `init` frames. It no longer treats `init` as a type, which means that if an init frame is not from core, it will now tell you so and treat it more appropriately. For example, here's a frame that is `init` and also from a dependency:

0x init

- More importantly, this puts us in a position where we're very nearly ready to add features based on additional types, such as checkboxes to filter out specific dependencies. We just need analysis to generate the list, which is WIP by @goto-bus-stop 

This touches on a lot of code so will need thorough testing for regressions.

### CODE REVIEW

**Comment by AlanSl** [comment]
Still needs tests, now works again with the new changes to master, particularly the new core / V8 categorisation. Here's another example of a rare init-but-not-core frame:

image

**Comment by AlanSl** [comment]
This PR's probably a good place to address this:

image

I'll add a commit setting the type for app frames to the app name so such things would be like `In profiled application (some-amazing-app)`

**Comment by goto-bus-stop** [comment]
there's a slightly strange colour variance here:

image

v8/v8 native have the grey colour, while node/v8 runtime have the greyish greenish one

**Comment by cuginoAle** [comment]
Late to the party... 
LGTM!

File: visualizer/ui.js [hunk 1]
```diff
@@ -120,7 +124,15 @@ class Ui extends events.EventEmitter {
   }
 
   selectHottestNode (opts) {
-    this.selectNode(this.dataTree.getFrameByRank(0), opts)
+    const node = this.dataTree.getFrameByRank(0)
+    const nodeInvalidMessage = ' node selected in selectHottestNode'
+
+    // Prevent infinite loop if some future bug allows an invalid node to be returned here
+    if (!node) throw new Error ('No' + nodeInvalidMessage)
```
Comment (cuginoAle) [inline visualizer/ui.js]: Little heads-up: You might have a Lint error here because of the space between `Error` and `('`

**Comment by cuginoAle** [comment]
It looks good to me.
One thing that would be [relevant for this](https://github.com/nearform/node-clinic-flame/issues/16) as well : we currently throw and error if `no node is selected`, that means we shouldn't be in a scenario where no data is displayed (either because the user filtered everything out or simply because 0x didn't collect any data).
Maybe we could notify the user that there's no data available  showing a `Message` rather then throwing an error?

File: visualizer/ui.js (hunk 1 above)
Comment (AlanSl) [reply visualizer/ui.js]: Strangely, the linter doesn't mind, but it probably _should_ mind, so I've changed it anyway 🙃

**Comment by AlanSl** [comment]
The difficulty with that issue is that there are lots of points in the process that assume that data exists and it's difficult to stop them all from hitting errors.

That comment has however given me an idea for a different way to solve the problem... and it seems to work. Will post a PR shortly, after merging this one.

**Pull request merged**